keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,732 @@
+
+```yaml
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+  name: disable-cross-namespace-affinity
+  namespace: foo-ns
+spec:
+  hard:
+    pods: "0"
+  scopeSelector:
+    matchExpressions:
+    - scopeName: CrossNamespaceAffinity
+```
+
+If operators want to disallow using `namespaces` and `namespaceSelector` by default, and 
+only allow it for specific namespaces,  they could configure `CrossNamespaceAffinity` 
+as a limited resource by setting the kube-apiserver flag --admission-control-config-file
+to the path of the following configuration file:
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: "ResourceQuota"
+  configuration:
+    apiVersion: apiserver.config.k8s.io/v1
+    kind: ResourceQuotaConfiguration
+    limitedResources:
+    - resource: pods
+      matchScopes:
+      - scopeName: CrossNamespaceAffinity
+```
+
+With the above configuration, pods can use `namespaces` and `namespaceSelector` only
+if the namespace where they are created have a resource quota object with 
+`CrossNamespaceAffinity` scope and a hard limit equal to the number of pods that are
+allowed to.
+
+Moreover, to prevent accidentally selecting a large number of namespaces, we will reject empty
+selectors. For example, users can do the following:
+
+```yaml
+podAntiAffinity:
+  requiredDuringSchedulingIgnoredDuringExecution:
+  - namespaceSelector:
+      matchExpressions:
+        - key: workload
+          operator: In
+          values:
+          - HPC
+```
+
+but can't do the following:
+
+```yaml
+podAntiAffinity:
+  requiredDuringSchedulingIgnoredDuringExecution:
+  - namespaceSelector: {}
+```
+
+For more protection, admission webhooks like gatekeeper can be used to further
+restrict the use of this field.
+
+#### External Dependencies
+We are aware of two k8s projects that could be impacted by this change, the descheduler 
+and cluster autoscaler. Cluster autoscaler should automatically consume the change
+since it imports the scheduler code, the descheduler however doesn't and needs to 
+be changed to take this feature into account. We will open an issue to inform the project
+about this update.
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+ Add `NamespaceSelector` field to `PodAffinityTerm`:
+
+```go
+type PodAffinityTerm struct {
+    // A label query over the set of namespaces that the term applies to.
+	// The term is applied to the union of the namespaces selected by this field
+	// and the ones listed in the namespaces field.
+	// nil selector and nil or empty namespaces list means "this pod's namespace"
```

Comment:
nit: `a nil selector and empty namespaces list means "this pod's namespace` (both must be unset)

Reply:
updated.